src/lib.rs

```diff
@@ -1,101 +1,405 @@
+                violations.lock().unwrap().extend(reported_violations);
+            }
+            if let Some(fixes) = query_match_context.into_pending_fixes() {
+                pending_fixes
+                    .lock()
+                    .unwrap()
+                    .entry(rule.meta.name.clone())
+                    .or_default()
+                    .extend(fixes);
+            }
+        },
+    )
+    .unwrap();
+    let mut violations = violations.into_inner().unwrap();
+    let pending_fixes = pending_fixes.into_inner().unwrap();
+    if pending_fixes.is_empty() {
+        return violations;
+    }
+    run_fixing_loop(
+        &mut violations,
+        file_contents,
+        pending_fixes,
+        tree_sitter_grep_args,
+        &aggregated_queries,
+        path,
+        &config,
+    );
+    violations
+}
+
+fn get_tree_sitter_grep_args(aggregated_queries: &AggregatedQueries) -> tree_sitter_grep::Args {
+    tree_sitter_grep::Args::parse_from([
         "tree_sitter_grep",
         "-q",
         &aggregated_queries.query_text,
         "-l",
         "rust",
         "--capture",
         CAPTURE_NAME_FOR_TREE_SITTER_GREP,
-    ]);
-    let reported_any_violations = AtomicBool::new(false);
-    tree_sitter_grep::run_with_callback(
-        tree_sitter_grep_args,
-        |capture_info, file_contents, path| {
-            let (rule_index, rule_listener_index) =
-                aggregated_queries.pattern_index_lookup[capture_info.pattern_index];
-            let rule = &resolved_rules[rule_index];
-            let listener = &rule.listeners[rule_listener_index];
-            (listener.on_query_match)(
-                capture_info.node,
-                &QueryMatchContext::new(
-                    path,
-                    file_contents,
-                    rule,
-                    &reported_any_violations,
-                    &context,
-                ),
-            );
+    ])
+}
+
+fn write_files<'a>(files_to_write: impl Iterator<Item = (&'a Path, &'a [u8])>) {
+    for (path, file_contents) in files_to_write {
+        fs::write(path, file_contents).unwrap();
+    }
+}
+
+type RuleName = String;
+
+fn apply_fixes(file_contents: &mut Vec<u8>, pending_fixes: HashMap<RuleName, Vec<PendingFix>>) {
+    let non_conflicting_sorted_pending_fixes =
+        get_sorted_non_conflicting_pending_fixes(pending_fixes);
+    for PendingFix { range, replacement } in non_conflicting_sorted_pending_fixes.into_iter().rev()
+    {
+        file_contents.splice(range, replacement.into_bytes());
+    }
+}
+
+fn compare_pending_fixes(a: &PendingFix, b: &PendingFix) -> Ordering {
+    if a.range.start < b.range.start {
+        return Ordering::Less;
+    }
+    if a.range.start > b.range.start {
+        return Ordering::Greater;
+    }
+    if a.range.end < b.range.end {
+        return Ordering::Less;
+    }
+    if a.range.end > b.range.end {
+        return Ordering::Greater;
+    }
+    Ordering::Equal
+}
```

Comment:
Could this be a `PartialOrd` implementation?

Reply:
(Which actually seems like it should delegate to a `PartialOrd` impl on the `range`.)

Reply:
It does seem like this is really just asserting a certain "ordering over `range`'s"

I think we'd run into the "orphan rule" in trying to implement `PartialOrd` for `ops::Range<usize>`

But that could work implementing it on "our" `PendingFix` type, added an issue